### ISSUE TITLE
Virtual call optimizations

### ISSUE THREAD

**Issue by jemc** [savi-lang/savi#266]
There are a variety of performance optimizations we can make to the virtual call system, which are related enough that they can probably be tackled together:

- Avoid including a method in the virtual table unless a call site for that method was reached that:
  - Calls it via an abstract type like a `:trait`
  - Calls it via a union type (e.g. calling `ptr` on a `(Bytes'box | String'box)` variable)
  - Calls it as an asynchronous behavior (all actor message numbers need the virtual table)
  
- Automatic specialization of functions (in `--release` mode only, since it's likely to be expensive)
  - When argument types of a call site are more specific than the parameter types,
    - Create an internal pseudo-reification of that function with the more specific parameter types
    - Continuing an analysis similar to today's `reach` analysis, but with the more specific pseudo-reifications
  - Each variable and each call site should be marked with its new more specific pseudo-type
    - This analysis will be consumed by CodeGen to create more specific LLVM code,
    - which can hopefully avoid a lot of virtual calls and avoid "boxing" a lot of `:struct`/`:numeric` values

**Comment by jasoncarr0** [comment]
One solution to prevent code blowup on Bullet 1 is via user input. Not perfect but GHC uses `{-# SPECIALIZE term :: type #-}` pragmas to allow the user to ensure specialization in certain cases.

**Comment by jemc** [comment]
Yes, that's a good callout. It's something I have thought about before but forget to mention here in the ticket.

We should have some kind of heuristic/threshold indicating when it's expected to be better to specialize vs not doing so. Perhaps allowing the threshold to be tunable at compile time.

And then as you mention, we should have a way to declare that a certain critical path *needs* to be fully specialized or fail at compile time if that's not possible.

**Comment by jemc** [comment]
Today I added another class of optimization to the ticket description: "Automatic elision of single-concrete abstract types"
